### ISSUE TITLE
Dealing with inherited and partial types

### ISSUE THREAD

**Issue by EndOfTheSpline** [Newex/PartialSourceGen#14]
It can be useful sometimes to re-use DTOs. You may have something that you need less information for e.g. a list/overview kinda scenario, which is a subset of your full data set (e.g. for the detail view).

As pseudo-code-ish example:

```cs
[Partial]
public record BookListItem(long Id, string Title, string Author, int Rating);

[Partial]
public record BookDetails(string Description, Image[] Images, Review[] Reviews): BookOverview;
```

Currently, `SemanticTransform` will only include the type that was `[Partial]`, i.e. only the properties declared directly on the type will be included in the partial type. This is actually kind of led me to another, very related issue too: Because it currently operates on the syntax of whatever the attribute is on, it does not work with partial [Partial]'s (which I find quite amusing 😄)

I think it would be neat if it could include base types too, but there's a few scenarios that might be tricky:

1. What happens in a scenario like above, where two types are both partial? Does `PartialBookDetails` inherit `PartialBookListItem`, therefore only defines its own declared properties (kinda similar to what it is doing today)?
2. What happens if the inherited type is `[Partial]`, but the base type is not? Does the Partial type declare all base properties as well?
3. What happens in a mixed scenario, if you have something like `[Partial]Third : Second : [Partial]Base`, i.e. a sandwich with a non-partial class in the middle?

I think a possible solution would be a recursive lookup of some kind, starting with the currently `[Partial]`'d type as TBase, and initializing TCurrent with TBase:

1. Include all properties that are declared on TCurrent
2. Set TCurrent = TCurrent.BaseType
3. If TCurrent is System.Object, break the recursion
4. If TCurrent is a Partial class, make TBase inherit from T, and break the recursion
5. Goto 1

This should allow for decent flexibility (covering all three scenarios above).

I'm currently tinkering a bit with it, and it seems like I can iterate the types and add the properties easily enough:

```cs
var props = new List<PropertyDeclarationSyntax>();
INamedSymbol? baseType = null;
for (var currentType = nameSymbol; currentType != null; currentType = currentType.BaseType)
{
    // if (currentType != nameSymbol && currentType.HasPartialAttribute())
    // {
    //     baseType = currentType;
    //     break;
    // }

    var newProps = currentType.GetMembers()
        .OfType<IPropertySymbol>()
        .SelectMany(s => s.DeclaringSyntaxReferences)
        .Select(s => s.GetSyntax())
        .OfType<PropertyDeclarationSyntax>();
    props.AddRange(newProps);
}
```

This will

- include all properties of the `[Partial]`'d type, including those defined in other files in the case of `partial` types
- include all properties of all base types of the `[Partial]`'d one
- not branch off properly when it encounters a parent `[Partial]`
- does nothing with the `baseType` (i.e. does not emit it in the `PartialInfo`)

I think it might make more sense to move into the symbol realm (as done here) for the property gathering, rather than relying on declarations, because it also fixes the partial issue briefly mentioned above. However, I don't know what the technical implications are of this - in my local tests with my local projects, it seems to work OK.

**Comment by Newex** [comment]
I have just copied/pasted your code, and run the tests. Everything green. Also added 3 test cases.

- Partial classes should have source generation.
- Partial class with spread out definitions should have all property members generated.
- A deep inheritance, should flatten and keep property members.

This is included in version 0.0.17

**Comment by EndOfTheSpline** [comment]
Hey, thanks! Sorry I didn't reply before, it's been kinda busy here.

Just to add to this: I've been doing some backroom thinking and I haven't really come to a conclusion whether related partials should also inherit from one another (i.e. if you have `[Partial]` on both `Model` and `Entity`, if `PartialModel` should inherit `PartialEntity` and only declare the partial properties that Model declares itself). For my case, it's not really required as it's a one-way street (I'm creating partials and giving them to others; I don't ever expect a partial back, so polymorphism isn't of interest).
